### DISCUSSION TITLE
logging resource contention when logging request for multiprocess Pool, windows specific?

### DISCUSSION THREAD

**Discussion by lz-work** [fastapi/fastapi#11284] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
import uvicorn
from multiprocessing import Pool
from fastapi import APIRouter
import logging
from logging.handlers import TimedRotatingFileHandler
from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint
from starlette.requests import Request
from starlette.responses import Response
import time
import asyncio


app = FastAPI()


def f():
    return 0

class Executor:
    __instance = None
    _pool = None

    def __new__(cls):
        if cls.__instance is None:
            cls.__instance = super().__new__(cls)
            pass
        return cls.__instance
        pass

    def __init__(self) -> None:
        if Executor._pool is None:
            Executor._pool = Pool(processes=1, maxtasksperchild=5)
        pass

    def add_task(self):
        assert Executor._pool is not None
        ## function f is a dummy function, it really does not matter
        Executor._pool.apply_async(f)
    pass


class HTTPRequestLogger:
    logger = None

    @classmethod
    def logger_init(cls):
        if cls.logger is None:
            HTTPRequestLogger.logger = logging.getLogger("x-ample-logger")
            file_handler = TimedRotatingFileHandler(
                filename='test.log',
                when="S",
                backupCount=90)
            HTTPRequestLogger.logger.setLevel(logging.DEBUG)
            formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s')
            file_handler.setFormatter(formatter)
            HTTPRequestLogger.logger.addHandler(file_handler)
        pass

    def __init__(self) -> None:
        HTTPRequestLogger.logger_init()
        pass

    async def log_request(self, request, response, duration: float):
        # log the entries
        HTTPRequestLogger.logger.info(f"{request} {duration}")


router = APIRouter()


@router.post('/add_task')
def add_task():
    """
    simply add the task to task queue, does not wait for the execution

    Returns:
        _type_: _description_
    """
    executor = Executor()
    executor.add_task()
    return "fixed string"

@router.post('/normal')
def stuff():
    """
    do some stuff without multiprocess pool then its ok

    Returns:
        str: fixed str
    """
    return 'OK'


class RequestLoggingMiddleware(BaseHTTPMiddleware):
    _logger = HTTPRequestLogger()

    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -> Response:
        request_init = time.perf_counter()
        response = await call_next(request)
        request_processed = time.perf_counter()
        # log the request
        asyncio.ensure_future(self._logger.log_request(request,
                                                       response,
                                                       request_processed - request_init))
        return response
    pass


app.add_middleware(RequestLoggingMiddleware)
app.include_router(router)


if __name__ == '__main__':
    uvicorn.run(app=app,
                host='0.0.0.0',
                port=3344
                )
    pass
```

### Description

**The issue is about the middleware passing the Logging File handle to the process pool's process and the File Handle then never gets released,** so that the rotation can never be performed by main thread.

The code I provided can be run directly, and basically have 2 parts, bear with me. The first is the endpoint part, and the second is the request logging part.

## The endpoints

there are 2 endpoint the /add_task and /normal

1. add_task

what is special about this class is that it gets a singleton class of executor, which have a multiprocess pool as its class variable, and add a task to the job queue to the pool. (basically starts another python process)

2.  normal

its just what you expect of a request, does nothing and returns a string

## the logging

the aim of the logging is to log the processing duration of the request, and its achieved via the middleware, short version, clocks in when request comes in, clocks out when request are done, then do a asyncio.ensure_future to do the log, simple.

the logger is basically the TimedRotatingFileHandler

## The issues
![image](https://github.com/tiangolo/fastapi/assets/145107434/e16d62b6-f9af-4007-b21f-cc30600f8471)

the logging rotation does not work after add_task endpoint is called. And the file descriptor of the logging file is passed to the spawned python process, and never released, despite it is never needed, which is probably the reason for the rotation to stop working. 

**I've use the resource manager of windows and process monitor to isolate the resource claimed by the bunch of python process, and confirmed its indeed passed to the spawned python process and is held there.** But why, I mean if you are referring to python's documentation [in here](https://docs.python.org/3.9/library/multiprocessing.html?highlight=pool#contexts-and-start-methods) you can see that spawning windows process does not inherit the file descriptor of the parent process. So maybe its something to do with FastAPI like does it pass along something to the pool? Or am I missing something

And thanks for anyone who is patient enough to read this chuck of words.

### Operating System

Windows

### Operating System Details

windows 10 

### FastAPI Version

0.110.0

### Pydantic Version

2.6.4

### Python Version

3.9.13

### Additional Context

few things I tried to troubleshoot this issue. (Updated 24/3/18)

#### Deploy it in Linux docker environment

picked the python:3.10.13-alpine3.18 as the base image, copy and pasted the code, runs OK, the rotation works, do not get any unwanted behavior

**Used JMeter to stress test the service under linux docker environment, everything is working, no logging issue at least.** which leads me to think this is a platform specific issue.

#### Remove and await for the logging to complete before returning the request

logging exceptions gets back propagate to the middleware, then user gets a HTTP code 500, Internal server error

**Comment by jgould22** [comment]
I am not sure about the lower level function (ie not create task) but you might want to check that your logging call to `asyncio.ensure_future` is not being garbage collected before it is actually completed.

I would start by just awaiting that call and see what happens.

(personally I just lean on struct log for logging but I can see how that might seem a bit heavy weight here)

**Reply by lz-work** [reply]
Jordan, thanks for your reply.

Actually before and after I posted this 'issue', I've already tried a few things, and your suggestion is just one of them, lol. What will happen after removing the ensure_future is the logging exception back propagates to the request, so that from the RESTful API caller's view point they gets the code 500 Internal server error. Which does not solve this issue. 

Among the things I've tried, I've used the same code making it running in my linux docker environment, it seemed to be function as intended. No exception what so ever, and rotation does work.

I intended to find out the cause of this behavior in Windows platform, have a few hunch, gna shoot them down one by one. Ideas are welcome.

will update the discussion for what I have tried

**Reply by lz-work** [reply]
And yes integrate with the logging analysis stack such as ELK is planned but I havn got that far yet.

**Comment by YuriiMotov** [comment]
Let's close this for now to keep the discussions clean and organized.
If you’ve found a solution, please share it here.
If you still need help, feel free to let us know.
